src/utils/db/car_stream.rs

```diff
@@ -137,6 +140,68 @@ impl<ReaderT: AsyncBufRead> Stream for CarStream<ReaderT> {
     }
 }
 
+pin_project! {
+    pub struct CarWriter<W> {
+        #[pin]
+        inner: W,
+        buffer: BytesMut,
+    }
+}
+
+impl<W: AsyncWrite> CarWriter<W> {
+    pub fn new_carv1(roots: Vec<Cid>, writer: W) -> anyhow::Result<Self> {
+        let car_header = CarHeader { roots, version: 1 };
+
+        let mut header_uvi_frame = BytesMut::new();
+        UviBytes::default().encode(Bytes::from(to_vec(&car_header)?), &mut header_uvi_frame)?;
+
+        Ok(Self {
+            inner: writer,
+            buffer: header_uvi_frame,
+        })
+    }
+}
+
+impl<W: AsyncWrite> Sink<(Cid, Vec<u8>)> for CarWriter<W> {
+    type Error = io::Error;
+
+    fn poll_ready(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        let this = self.project();
+        if !this.buffer.is_empty() {
+            let bytes_written = ready!(this.inner.poll_write(cx, this.buffer))?;
+            this.buffer.advance(bytes_written);
+            if !this.buffer.is_empty() {
+                return Poll::Pending;
+            }
+        }
+        Poll::Ready(Ok(()))
+    }
+    fn start_send(self: Pin<&mut Self>, item: (Cid, Vec<u8>)) -> Result<(), Self::Error> {
+        let this = self.project();
+
+        let payload = [item.0.to_bytes(), item.1].concat();
+        let mut var_bytes = [0; 10];
+        let len = payload.len().encode_var(&mut var_bytes);
+
+        this.buffer.extend_from_slice(&var_bytes[0..len]);
+        this.buffer.extend_from_slice(&payload);
+
+        Ok(())
+    }
+    fn poll_flush(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        // TODO: find out if we really need to call poll_ready here?
```

Comment:
I think we have to call `poll_ready` here. Why do you think otherwise?

Reply:
Quoting doc for `poll_flush` method:

> Returns Poll::Ready(Ok(())) when no buffered items remain. If this value is returned then it is guaranteed that all previous values **sent** via start_send have been flushed.

My understanding is that an item is **sent** in a two-step way. First `start_send` is called, then encoded item is written via `poll_write`.
So consider the sequence of calls: `start_send item1`, `poll_ready item1 -> ready`, `start_send item2`, `poll_flush`, we know we can return `Poll::Ready(Ok(()))` for flush because `item1` has been fully sent (which is not the case for `item2` yet).

But since I'm not sure about this reasoning I will keep the call to `poll_ready`.